### PR TITLE
Modify AddressChangeRequest to use SlotRange

### PR DESCRIPTION
**Pull request by nau**
Modify `AddressChangeRequest` to use `SlotRange`. Now it returns all transactions that modify an address within the SlotRange. 
This modified `AddressChangeRequest` now can be used to get a history of a contract execution. This is needed for Marlowe Run contract execution in particular.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [x] Commits have useful messages
- [x] Review clarifications made it into the code
- [x] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-contract/src/Wallet/Emulator/ChainIndex/Index.hs [hunk 1]
```diff
@@ -102,12 +102,21 @@ insert am item (ChainIndex ci) =
     in ChainIndex (foldl' (\ci' addr -> Map.alter alt addr ci') ci keys)
 
 -- | All transactions that modify the address, from a given slot onwards
```
Comment (j-mueller) [inline plutus-contract/src/Wallet/Emulator/ChainIndex/Index.hs:104]:
```suggestion
-- | All transactions that modify the address in the interval
```

File: plutus-ledger-api/src/Plutus/V1/Ledger/Interval.hs [hunk 2]
```diff
@@ -71,11 +77,23 @@ data UpperBound a = UpperBound (Extended a) Closure
...
+    pretty (UpperBound NegInf _) = pretty "-∞)"
+    pretty (UpperBound a True)   = pretty a <+> pretty "]"
+    pretty (UpperBound a False)  = pretty a <+> pretty ")"
+
 -- | The lower bound of an interval.
 data LowerBound a = LowerBound (Extended a) Closure
     deriving stock (Haskell.Eq, Haskell.Ord, Show, Generic)
     deriving anyclass (FromJSON, ToJSON, Serialise, Hashable, NFData)
 
+instance Pretty a => Pretty (LowerBound a) where
```
Comment (j-mueller) [inline plutus-ledger-api/src/Plutus/V1/Ledger/Interval.hs:91]: 👍

File: plutus-contract/src/Wallet/Emulator/ChainIndex/Index.hs [hunk 3]
```diff
@@ -102,12 +102,21 @@ insert am item (ChainIndex ci) =
...
-    $ split sl
-    $ snd
-    $ split (pred sl)
-    $ Map.findWithDefault mempty addr mp
+transactionsAt :: ChainIndex -> SlotRange -> Address -> [ChainIndexItem]
+transactionsAt (ChainIndex mp) sl addr = let
+    allItems = Map.findWithDefault mempty addr mp
+    result = case sl of
+        Interval (LowerBound (Finite s1) in1) (UpperBound (Finite s2) in2) ->
+            let low = if in1 then s1 else pred s1
```
Comment (j-mueller) [inline plutus-contract/src/Wallet/Emulator/ChainIndex/Index.hs:110]: should this not be `else succ s1` ?

**Comment by michaelpj** [comment]
Can we afford to do this? This requires the whole history of the chain, and I'm not sure how we wanted to do address-watching...

**Comment by j-mueller** [comment]
True, that's now how we wanted to do it. There are some modifications planned for the chain index - I want to get rid of the ability to query by slot altogether.

This PR a short-term fix for a marlowe dashboard problem.

File: plutus-contract/src/Wallet/Emulator/ChainIndex/Index.hs (hunk 3 above)
Comment (nau) [reply plutus-contract/src/Wallet/Emulator/ChainIndex/Index.hs]: Please, take another look. I've fixed this a bit.

**Review by j-mueller — APPROVED** [review]

**Pull request merged**